diff --git a/fs/namespace.c b/fs/namespace.c
index 4d910be7c..ee663c30a 100644
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -1786,6 +1786,10 @@ static int can_umount(const struct path *path, int flags)
 	struct mount *mnt = real_mount(path->mnt);
 	struct super_block *sb = path->dentry->d_sb;
 
+	#ifdef CONFIG_KSU_SUSFS
+	if (susfs_is_current_proc_umounted())
+		return -EACCES;
+	#endif
 	if (!may_mount())
 		return -EPERM;
 	if (path->dentry != path->mnt->mnt_root)
