diff --git a/fs/open.c b/fs/open.c
index 82d10b7a5..3a7c9a329 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -33,6 +33,9 @@
 #include <linux/dnotify.h>
 #include <linux/compat.h>
 #include <linux/mnt_idmapping.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#endif
 
 #include "internal.h"
 #include <trace/hooks/syscall_check.h>
@@ -1334,6 +1337,11 @@ static long do_sys_openat2(int dfd, const char __user *filename,
 	if (fd)
 		return fd;
 
+	#ifdef CONFIG_KSU_SUSFS
+	if (susfs_check_unicode_bypass(filename))
+		return -ENOENT;
+	#endif
+
 	tmp = getname(filename);
 	if (IS_ERR(tmp))
 		return PTR_ERR(tmp);
