--- a/kernel/feature/sucompat.c	2026-09-18 18:41:16.687479580 -0600
+++ b/kernel/feature/sucompat.c	2026-09-18 18:41:16.687479580 -0600
@@ -88,6 +88,7 @@
 	return true;
 }
 
+#ifndef CONFIG_KSU_SUSFS
 long ksu_handle_faccessat_sucompat(int orig_nr, struct pt_regs *regs)
 {
 	const char __user **filename_user, *orig_filename;
@@ -123,6 +124,7 @@
 do_orig_facessat:
 	return ksu_syscall_table[orig_nr](regs);
 }
+#endif
 
 long ksu_handle_stat_sucompat(int orig_nr, struct pt_regs *regs)
 {
